editors/code/src/test_explorer/discover_and_update.ts

```diff
@@ -0,0 +1,594 @@
+    const runnables = testInfos.map(it => new RunnableFacde(it.runnable));
+
+    const testModuelRunnables = runnables.filter(it =>
+        it.testKind === NodeKind.TestModule)
+        .sort(RunnableFacde.sortByLabel);
+
+    const testItemRunnables = runnables.filter(it =>
+        it.testKind === NodeKind.Test);
+
+    assert(testModuelRunnables.length + testItemRunnables.length === runnables.length);
+
+    const rootTestModuleRunnbale = testModuelRunnables[0];
+
+    // Now, we know the root test module
+    // Note that the parent might be not exist in the tree.
+    // Then how to update the test model tree?
+    // There are two ways:
+    // - down-to-up, which means get parent module,  and create parent recursively
+    // - up-to-down, find the nearest parent, and create children for it
+    // up-to-down is simpler, since the logic could be reused,
+    // for anyway we need to init, which is up-to-dowm :P
+    // PS: we could set dirty flag to make it more effective, but I think it's fine
+
+    // Observe that:
+    // if node is new in the tree, fetch its children too
+    // if node is existing in the tree,
+    // - and it's definition, then we have got all of the runnables(because they
+    // are in same file) and will be updated
+    // - and it's declaration
+    //   - and definition uri is not changed, then not fetch children
+    //   - and definition uri is changed, then fetch the children
+
+    // So, the final steps are:
+    // 1. Make the change(remove, add new test module declaration or add test module with items),
+    // but not fetch children for new test module declaration
+    // 2. collectt which children need to be fetched(reuse logic)
+    // 3. fetch children(reuse logic)
+
+    const nearestNode = testModelTree.findNearestNodeByRunnable(rootTestModuleRunnbale);
+
+    assert(nearestNode.kind !== NodeKind.Test, "the nearest node of root module can not be a test node.");
+
+    // create target node
+    // This is necessary, because we do not know how many targets a package contains unless we fetch data of `cargo metadata`
+    // But we want to only fetch it when cargo file is changed, to make things more lazily.
+    if (rootTestModuleRunnbale.origin.label === "test-mod "
+        && nearestNode.kind !== NodeKind.TestModule) {
+        assert(nearestNode.kind === NodeKind.CargoPackage);
+        // TODO: what should we do if user add a new package or workspace?
+        // Maybe listen to the change of cargo file, and always refresh everything
+        // This runnable is from a target, create the target if it's not exist in test model tree
+        const parentLocation = await RaApiHelper.parentModue(rootTestModuleRunnbale.uri);
+        assert(parentLocation?.length === 1, "The changed file contains runnables, there must be parent modules, and it must be cargo file.");
+        const targetUri = parentLocation[0].targetUri;
+        assert(targetUri.toLowerCase().endsWith("cargo.toml"));
```

Comment:
the project manifest is not necessarily guaranteed to a Cargo.toml. I'd consider making use of the manifest path in https://github.com/rust-lang/rust-analyzer/pull/14931 if/when it lands.